### ISSUE TITLE
@var MyTrait $mockForTrait causes UndefinedDocblockClass

### ISSUE THREAD

**Issue by caugner** [vimeo/psalm#3695]
When testing traits using phpUnits' `TestCase::getMockForTrait()` I would like to have IDE autocompletions.

However, type hinting the mock with the Trait "class" causes Psalm to complain:

https://psalm.dev/r/8d489615b1

**Comment by caugner** [comment]
Also, `@psalm-suppress`ing only suppresses the error for the initial type hint, not for the method call (but maybe this is intended?):

https://psalm.dev/r/ba5df0544b

**Comment by sanmai** [comment]
Another example: https://psalm.dev/r/b033c99c11

**Comment by Radiergummi** [comment]
I have another use case for this; in Laravel, Notification classes will generally be invoked with an instance of a model that [uses the `Notifiable` trait](https://laravel.com/docs/8.x/notifications#using-the-notifiable-trait), like so:
```php
class MyNotification extends Notification
{
    public function toMail(Notifiable $recipient): MailMessage
    {
        // ...
    }
}
```

Alas, `Notifiable` is not an interface, and there is none for it. I'm not a huge Laravel fan, but that's what we have to deal with anyway.  
Due to this issue, the parameter cannot be type-hinted accurately.

**Comment by orklah** [comment]
This issue is emitted here: https://github.com/vimeo/psalm/blob/b29737f2b154748430250effdbcf47de8a5904df/src/Psalm/Internal/Analyzer/ClassLikeAnalyzer.php#L285

If you look a few lines above, there's this: https://github.com/vimeo/psalm/blob/b29737f2b154748430250effdbcf47de8a5904df/src/Psalm/Internal/Analyzer/ClassLikeAnalyzer.php#L282

who looks like it could allow traits if a specific flag is true. If someone wants to try to make progress on this, I guess a good start would be to check try to toggle this flag to true and check if this could work.

I won't work on this because I don't use traits much and I'm very not motivated to work with them, but I'll offer help if I can to anyone who wants to try

**Comment by ricardoboss** [comment]
I experimented a bit in https://github.com/ricardoboss/psalm/tree/issue/3695-trait-in-docblock-error and it seems to be more complex than I initially thought. The only changes include allowing traits when checking for a class like type in a docblock and for a method call target (i.e. what @orklah proposed).

A workaround is to just give Psalm what it wants: an interface to check the method calls against: https://psalm.dev/r/8067be27b9

**Comment by orklah** [comment]
I tried myself and it comes to: https://github.com/orklah/psalm/tree/traits2

There's one failure in tests:
https://psalm.dev/r/72d6bd8836

After the changes, this doesn't emit an error anymore. Does anyone know why this code is bad?

**Comment by ricardoboss** [comment]
@orklah it is bad because then you would be able to use traits as property types, as the testsuites indicate: https://github.com/orklah/psalm/runs/4292943773?check_suite_focus=true & https://github.com/orklah/psalm/runs/4292945923?check_suite_focus=true#step:7:241

**Comment by weirdan** [comment]
Traits are not types, they do not provide any guaranties as their methods may be aliased, shadowed by another trait and their visibility can be changed.

**Comment by orklah** [comment]
@weirdan so, by that logic, the original example should stay an error and Psalm is just doing its job?

**Comment by weirdan** [comment]
That would be my take, yes.

**Comment by orklah** [comment]
Okay, I'll let this open for now, it seem to be a popular issue. We should at least try to formalize this in the documentation before closing. I'll change the labels though.

**Comment by weirdan** [comment]
To provide a concrete example, the `C` class in the following snippet does not implement neither T1 nor T2 'interfaces': https://3v4l.org/soWld

**Comment by ricardoboss** [comment]
After checking if an object uses a specific trait ([`class_uses`](https://www.php.net/manual/en/function.class-uses.php)) we should be able to give Psalm some kind of type hint.

I imagine something like an intersection type could work: https://psalm.dev/r/77767bde76

**Comment by weirdan** [comment]
`&T1` (if `T1` is treated as a type) implies `foo()` will return `int` _and_ `bar()` will return `string`.

**Comment by ricardoboss** [comment]
Correct, but in a perfect world, the whole docblock would be unnecessary. Maybe the `@var` annotation isn't sufficient for declaring what method signature of a specific trait should be used for type checking...
